File: python-version/group.py
from typing import List

from constants import DX, DY, EMPTY
from point import Point


def is_in_board(x: int, y: int, size: int) -> bool:
    return 1 <= x <= size and 1 <= y <= size


class Group:
    def __init__(self, x: int, y: int, size: int):
        self.liberties = 0
        self.length = 1
        self.stones = set()
        self.st = [[False for _ in range(size + 1)] for _ in range(size + 1)]
        self.reset(size)
        self.add_to_group(x, y)

    def add_to_group(self, x: int, y: int):
        self.stones.add(Point(x, y))

    def reset(self, size: int):
        for x in range(1, size + 1):
            for y in range(1, size + 1):
                self.st[x][y] = False

    def get_group_length(self, x: int, y: int, color: int, size: int, board: List[List[int]]):
        for i in range(4):
            nx, ny = x + DX[i], y + DY[i]
            if not is_in_board(nx, ny, size) or self.st[nx][ny]:
                continue
            if board[nx][ny] == EMPTY:
                self.liberties += 1
                self.st[nx][ny] = True
                continue
            if board[nx][ny] != color:
                self.st[nx][ny] = True
                continue
            self.st[nx][ny] = True
            self.length += 1
            self.add_to_group(nx, ny)
            self.get_group_length(nx, ny, color, size, board)

    def get_group_length_and_liberty(self, x: int, y: int, color: int, size: int, board: List[List[int]]):
        self.reset(size)
        self.get_group_length(x, y, color, size, board)

File: java-version/src/main/java/org/sdu/go/entity/Group.java
package org.sdu.go.entity;

import org.sdu.go.common.Constants;

import java.util.HashSet;
import java.util.Set;

public class Group {

    private int liberties;
    private int length;
    public Set<Point> stones;
    private final Boolean[][] st;

    public Group(int x, int y, int size) {
        this.liberties = 0;
        this.length = 1;
        stones = new HashSet<>();
        st = new Boolean[size + 1][size + 1];
        reset(size);
        add2Group(x, y);
    }

    public int getLiberties() {
        return liberties;
    }

    public int getLength() {
        return length;
    }

    private void add2Group(int x, int y) {
        Point point = new Point(x, y);
        stones.add(point);
    }

    private void reset(int size) {
        for (int x = 1; x <= size; x++) {
            for (int y = 1; y <= size; y++) {
                st[x][y] = Boolean.FALSE;
            }
        }
    }

    public boolean isInBoard(int x, int y, int size) {
        return (x > 0 && x <= size && y > 0 && y <= size);
    }

    /***
     * 从一个位置开始 递归查找与该棋子相邻的同色所有棋子
     * @param x 位置横坐标
     * @param y 位置纵坐标
     * @param color 黑棋或白棋
     */
    private void getGroupLength(int x, int y, int color, int size, int[][] board) {
        for (int i = 0; i < 4; i ++ ) {
            int nx = x + Constants.dx[i], ny = y + Constants.dy[i];
            if (!isInBoard(nx, ny, size) || st[nx][ny]) continue;
            if (board[nx][ny] == Constants.EMPTY) {
                this.liberties ++;
                st[nx][ny] = Boolean.TRUE;
                continue;
            }
            if (board[nx][ny] != color) {
                st[nx][ny] = Boolean.TRUE;
                continue;
            }
            st[nx][ny] = Boolean.TRUE;
            this.length ++;
            add2Group(nx, ny);
            // 递归判断
            getGroupLength(nx, ny, color, size, board);
        }
    }

    // 从一个点开始 遍历从这个点延伸出去的组的长度和气
    public void getGroupLengthAndLiberty(int x, int y, int color, int size, int[][] board) {
        reset(size);
        getGroupLength(x, y, color, size, board);
    }
}
